### ISSUE TITLE
7.4-Transformer Training

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#48]
剛才講的都還只是假設你模型訓練好以後它是怎麼運作的, 接下來就要講怎麼做訓練 

如果是做語音辨識, 那你要有訓練資料
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/c2a5e71c-9e92-4442-b0a8-7e7b9fa1d51b)

你要蒐集一大堆的聲音訊號
每一句聲音訊號都要有工讀生來聽, 並Label出它的這個對應的詞彙是什麼

工讀生聽這段是機器學習, 他就把機器學習四個字打出來

所以Model就知道說你的這個 Transformer 應該要學到 聽到這段聲音訊號
它的輸出就是機器學習這四個中文字

### 如何讓機器學到這件事呢?
我們已經知道說輸入這段聲音訊號, 第一個應該要輸出的中文字是機
所以今天當我們把 BEGIN 丟給這個 Encoder 的時候, 它第一個輸出應該要跟機越接近越好

輸出的這個字會被表示成一個 One-Hot 的 Vector
在這個 Vector 裡面 只有機對應的那個維度是 1 其他都是 0 這是正確答案
Decoder 它的輸出是一個 Distribution 是一個機率的分布

我們會希望這一個機率的分布跟這個 One-Hot 的 Vector 越接近越好
所以去計算這個 Ground Truth跟這個 Distribution 它們之間的 Cross Entropy 
然後希望這個 Cross Entropy 的值越小越好(跟分類很像)

你可以想成每一次 Decoder 在產生一個中文字的時候其實就是做了一次分類的問題
中文字假設有四千個, 那就是做有四千個類別的分類的問題
我們已經知道輸出應該是機器學習這四個字
那你就告訴你的 Decoder 說
現在你第一次的輸出到第四次輸出應該分別就是機 器 學跟習這四個中文字的 One-Hot Vector
我們希望我們的輸出 跟這四個字的 One-Hot Vector 越接近越好
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/78e9f962-51a8-4941-ab88-283b654478ee)

在訓練的時候每一個輸出跟它對應的正確答案( One-Hot Vector)都有一個 Cross Entropy
我們要希望所有的 Cross Entropy 的總和越小越好

不要忘了還有 END 這個東西
所以其實今天假設這個句子它的中文的字是四個
但是你學習的時候 你要 Decoder 輸出的不是只有這四個中文字, 還要記得輸出END 這個特別的符號
所以你要告訴你的 Decoder 說 你最終第五個位置輸出了這個向量應該跟END 的 One-Hot Vector

### Teacher Forcing

在訓練的時候我們會給 Decoder 看正確答案=我們會在輸入的時候給它正確的答案

也就是我們會告訴它說

有 BEGIN 有機的情況下要輸出器
有 BEGIN 有機 有器的情況下輸出學
有 BEGIN 有機 有器 有學的情況下輸出習
有 BEGIN 有機 有器 有學 有習的情況下就要輸出END 

也就是我們把正確的答案 當作 Decoder 的輸入

但是測試的時候顯然沒有正確答案可以給 Decoder 看, 在 Influence 的時候 Decoder 看到的是自己的輸入
這中間顯然很容易有一個 Mismatch(之後會說明解決方式)

__________

# 訓練這種 Sequence To Sequence Model (包含Transformer) 的 Tip

### Copy Mechanism

在我們剛才的討論裡面, 我們都要求 Decoder 自己產生輸出
但是對很多任務而言, 也許 Decoder 沒有必要自己創造輸出出來
它需要做的事情 也許是從輸入的東西裡面複製一些東西出來

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/221919fe-105f-49fe-b7f2-dae5abf75635)

哪些任務會用得上?
聊天機器人 舉例來說 人對機器說 你好 我是庫洛洛
機器應該回答說 庫洛洛你好 很高興認識你
那對機器來說, 它其實沒有必要創造庫洛洛這個詞彙(這對機器來說一定會是一個非常怪異的詞彙)
因為在訓練資料裡面可能一次也沒有出現過 所以它不太可能正確地產生這段詞彙出來

但是假設今天機器它在學的時候
它學到的並不是它要產生庫洛洛這三個中文字
它學到的是看到輸入的時候說我是某某某 就直接把某某某
不管這邊是什麼複製出來說某某某你好, 那這樣子機器的訓練顯然會比較容易
它顯然比較有可能得到正確的結果 

所以複製對於對話來說可能是一個需要的技術需要的能力

另外一個例子-摘要, 但就不細講了

最早有從輸入複製東西的能力的模型叫做 Pointer Network 後來變形 叫做 Copy Network(可以看一下 Copy Mechanism)
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/70fdf465-fa53-4e1a-ae12-2089916e405c)

__________

### Guided Attention 

機器它會犯非常低級的錯誤
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/3c778f7d-5aaa-4bf5-a221-f6e4782cf253)

我輸入的發財是明明是同樣的詞彙 只是重複 4 次機器居然自己有一些抑揚頓挫 
它怎麼學到這件事 ?不知道, 機器它自己訓練出來就是這個樣子 

然而讓它講 1 次發財, 發現沒有唸發 
就是你這個 Sequence-To-Sequence Model 有時候 Train 出來就是會產生莫名其妙的結果

也許在訓練資料裡面這種非常短的句子很少
不知道要怎麼處理這種非常短的句子
要花很多時間才找得到這種差的例子 但這樣子的例子是存在的

Guided Attention = 強迫機器一定要把輸入的每一個東西通通看過
它最適合的是 像這種語音辨識 語音合成這種任務

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/0464e2a3-c4ee-4721-8d6d-409cbb9a8c1b)

因為像語音辨識這種任務 你其實很難接受說 你講一句話今天沒有辨識出來
或語音合成你輸入一段文字 語音合出來居然有一段沒有唸到
這個人很難接受, 但如果是其它應用 比如說 Chat Bar 或者是 Summary 可能就沒有那麼嚴格

Guiding Attention 要做的事情就是我們要去這個 Guide機器, 去領導這個 Attention的過程
要求機器它在做 Attention 的時候是有固定的方式的

當你發現機器的 Attention 是顛三倒四的
它先看最後面接下來再看前面, 那顯然這樣子的 Attention 是有問題的

Guiding Attention 要做的事情就是 強迫 Attention 有一個固定的樣貌 

語音合成 TTS 這樣的問題 你的 Attention 的分數 Attention 的位置都應該由左向右
那不如就直接把這個限制放進你的 Training 裡面, 要求機器學到 Attention就應該要由左向右

關鍵詞彙 Mnotonic Attention 或 Location-Aware 的 Attention (大坑不細講)

### Beam Search

假設 Decoder 它就只能產生兩個字一個叫做 A 一個叫做 B (假如世界上只有兩個字 A 跟 B)
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/8e89028c-d751-4c9c-b578-06507b5be15c)

那對 Decoder 而 它做的事情就是每一次Time Step 在 A B 裡面決定一個

然後決定了 A 以後再把 A 當做輸入然後再決定 A B 要選哪一個

那在我們剛才講的 Process 裡面每一次 Decoder 都是選分數最高的那一個

假設 A 的分數 0.6 B 的分數 0.4 Decoder 的第一次就會輸出 A
然後接下來假設 B 的分數 0.6 A 的分數 0.4 Decoder 就會輸出 B 
然後再假設把 B 當做 Input 就現在輸入已經有 A 有 B 了 然後接下來
A 的分數 0.4 B 的分數 0.6 那 Decoder 就會選擇輸出 B 所以輸出就是 A 跟 B 跟 B
那像這樣子每次找分數最高的那個 Token 每次找分數最高的那個字
來當做輸出這件事情叫做 Greedy Decoding

但是 Greedy Decoding 不一定是好的方法

比如說第一步雖然 B 是 0.4, 但接下來的 B 的可能性就大增變成 0.9 

如果比較紅色的這一條路跟綠色這條路的話
你會發現說綠色這一條路, 雖然一開始第一個步驟你選了一個比較差的輸出, 但是接下來的結果是好的

但問題是我們實際上並沒有辦法爆搜所有可能的路徑
因為實際上每一個轉捩點可以的選擇太多了
我們中文有 4000 個字 所以這個樹每一個地方分叉都是 4000 個可能的路徑

你走兩三步以後 你就無法窮舉->有一個演算法叫做 Beam Search

Beam Search它用比較有效的方法找一個估測的 Solution
它有時候有用 有時候沒有用

你會看到有些文獻告訴你說Beam Search 是一個很爛的東西

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/91c67c0f-4c5d-4770-9609-4b2638a5a62b)

The Curious Case Of Neural Text Degeneration 

那這個任務要做的事情是Sentence Completion 
也就是機器先讀一段句子, 接下來它要把這個句子的後半段把它完成 
你給它一則新聞, 或者是一個故事的前半部, 它自己發揮它的想像創造力把這個文章故事的後半部把它寫完 

你會發現說Beam Search 在這篇文章裡面一開頭就告訴你說 Beam Search 自己有問題
如果用 如果你用 Beam Search 的話 會發現說機器不斷講重複的話

它不斷開始陷入鬼打牆 無窮迴圈 不斷說重複的話

那如果你今天不是用 Beam Search有加一些隨機性 雖然結果不一定完全好 但是看起來至少是比較正常的句子
所以有趣的事情是有時候對 Decorder 來說 沒有找出分數最高的路反而結果是比較好的

Q:要找出分數最高的路又講說找出分數最高的路不見得比較好??
A:看你的任務的本身的特性

假設一個任務它的答案非常地明確:比如說語音辨識通常 Beam Search 就會比較有幫助
說一句話辨識的結果就只有一個可能

需要機器發揮一點創造力的時候這時候 Beam Search 就比較沒有幫助

__________________

# BLEU Score

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/6c4c1c5c-9ae9-4bc1-8689-d6c9e520a877)

BLEU Score 是你的 Decoder 先產生一個完整的句子以後再去跟正確的答案一整句做比較
我們是拿兩個句子之間做比較才算出 BLEU Score

但我們在訓練的時候顯然不是這樣
訓練的時候我們 Minimize 的是 Cross Entropy (每一個詞彙是分開考慮的)

但是不一定可以 Maximize BLEU Score 

所以你發現說在助教的程式裡面
在做 Validation 的時候並不是拿 Cross Entropy最低來挑最好的 Model 
而是挑 BLEU Score 最高的那一個 Model 

```
所以我們訓練的時候是看 Cross Entropy 但是我們實際上你作業真正評估的時候 看的是 BLEU Score
```

Q:那我們能不能在 Training 的時候 就考慮 BLEU Score 
A:但是 BLEU Score 本身很複雜它是不能微分的, 你把它當做你的 Loss 你根本不知道你要怎麼算

但是當遇到你在 Optimization 無法解決的問題, 用 RL 硬 Train 一發就對了這樣
當然這是一個比較難的做法, 並沒有特別推薦你在作業裡面用這一招

______________________

# Scheduled Sampling 避免 exposure bias

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/140dc925-f453-4eba-996c-812e9b5aed09)

exposure bias = 訓練跟測試不一致的現象

在訓練的時候 Decoder 看到的是完全正確的解答
但是測試的時候 Decoder 看到的是自己的輸出, 所以測試的時候 Decoder 會看到一些錯誤的東西

那假設 Decoder 在訓練的時候永遠只看過正確的東西
那在測試的時候你只要有一個錯,那就會一步錯,步步錯
因為對 Decoder 來說它從來沒有看過錯的東西
它看到錯的東西會非常的驚奇, 然後接下來它產生的結果可能都會錯掉

Q:所以要怎麼解決exposure bias的問題呢
A:Scheduled Sampling
Scheduled Sampling的方法就是給 Decoder 的輸入加一些錯誤的東西
你不要給 Decoder 都是正確的答案, 偶爾給它一些錯的東西它反而會學得更好
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/c79a3de2-0744-4603-836a-19a35d950e4f)

但是 Scheduled Sampling 這一招 它其實會傷害到 Transformer 的平行化的能力
